### PR TITLE
Sort binary

### LINKED ISSUE #568: Implement sort() for binary array
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
Implement `sort()` function for `BinaryArray`, `LargeBinaryArray` and `FixedSizeBinaryArray`.

**Describe the solution you'd like**
Extend "sort" and "take" kernels to binary array family.

**Describe alternatives you've considered**
n/a

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by waynexia**
# Which issue does this PR close?

Closes #568.

# What changes are included in this PR?

- Extend take kernel to binary array family
- Implement sort for binary array family
- Unit test

### CODE REVIEW

File: arrow/src/compute/kernels/sort.rs [hunk 1]
```diff
@@ -2375,6 +2494,179 @@ mod tests {
         );
     }
 
+    #[test]
+    fn test_sort_binary() {
+        test_sort_binary_arrays(
```
Comment (alamb) [inline arrow/src/compute/kernels/sort.rs:2499]: May I suggest at least one test where the fixed size vector has a non zero value in index `0` or `1`? Like `vec[1, 0, 3]` or something to demonstrate that the entire array is being checked?

**Review by alamb — COMMENTED** [review]
Thank you @waynexia 

I am not super familiar with the internals of list generic binary array, but I reviewed the tests you have added (thank you!) and they seem to demonstrate the correct behavior. 

@nevi-me  or @jorgecarleitao  might you have some time to review this PR?

File: arrow/src/compute/kernels/sort.rs [hunk 2]
```diff
@@ -2545,6 +2545,31 @@ mod tests {
             Some(3),
         );
 
+        test_sort_binary_arrays(
```
Comment (alamb) [inline arrow/src/compute/kernels/sort.rs:2548]: ❤️

File: arrow/src/compute/kernels/sort.rs (hunk 1 above)
Comment (waynexia) [reply arrow/src/compute/kernels/sort.rs:2499]: Thanks! This is a good point. I add a case with different prefix in 708e99b to cover.

**Comment by alamb** [comment]
Unless I hear anything else I plan to merge this later today or tomorrow

**Comment by alamb** [comment]
Thanks again @waynexia 🚀

**Comment by waynexia** [comment]
Appreciate it @alamb!

**Comment by alamb** [comment]
BTW @waynexia I expect this change to be available in arrow 5.1.0 (scheduled for release at the end of this week or early next week)

**Comment by waynexia** [comment]
Sure! I'd like to help. 

It seems that changes in #552 make the cherry-pick PR's CI failed. I think #552 should be cherry-picked first. 

However, #552 changes the default behavior (from stable sort to unstable version). So if ship #552 in a minor release is not expected, I can submit a patch to make #621 pass CI.

**Comment by waynexia** [comment]
Ahh, I noticed #573 which already gives a try.

May I ask which way should I add changes to #621? Add commits to branch `cherry_pick_2113e2b1` (if I can) or file a new PR.

**Comment by alamb** [comment]
Thank you very much @waynexia ! 

> May I ask which way should I add changes to #621? Add commits to branch cherry_pick_2113e2b1 (if I can) or file a new PR.

I think a new PR would be best (feel free to base it off of `cherry_pick_2113e2b1`) so that it is clear what is happening

**Comment by waynexia** [comment]
Got it, I opened #624 :smiley:

**Pull request merged**